core/frontend/src/render/webgl/AtmosphereUniforms.ts

```diff
@@ -0,0 +1,218 @@
+    const earthPolarRadius = this._earthScaleMatrix.at(2, 2);
+    const minDensityThresholdRadius = earthPolarRadius + atmosphereHeightAboveEarth;
+    this._atmosphereRadiusScaleFactor = (earthPolarRadius === 0)
+      ? 1
+      : (minDensityThresholdRadius / earthPolarRadius);
+  }
+
+  private _updateAtmosphereMaxDensityThresholdScaleFactor(maxDensityDepthBelowEarth: number) {
+    const earthPolarRadius = this._earthScaleMatrix.at(2, 2);
+    const maxDensityThresholdRadius = earthPolarRadius - maxDensityDepthBelowEarth;
+    this._atmosphereMaxDensityThresholdScaleFactor = (earthPolarRadius === 0)
+      ? 1
+      : (maxDensityThresholdRadius / earthPolarRadius);
+  }
+
+  private _updateDensityFalloff(densityFalloff: number) {
+    this._densityFalloff = densityFalloff;
+  }
+
+  private _updateScatteringCoefficients(scatteringStrength: number, wavelengths: Atmosphere.Wavelengths) {
+    // Rayleigh scattering strength is inversely related to the 4th power of the wavelength -> 1/pow(wavelength, 4)
+    // Because this produces very small values when the wavelengths are taken in nanometers,
+    //   we attempt to normalize them around 1 by taking the smallest wavelength of visible light as a baseline (violet light - 400nm)
+    const violetLightWavelength = 400.0;
+    this._scatteringCoefficients[0] = ((violetLightWavelength / wavelengths.r) ** 4.0) * scatteringStrength;
+    this._scatteringCoefficients[1] = ((violetLightWavelength / wavelengths.g) ** 4.0) * scatteringStrength;
+    this._scatteringCoefficients[2] = ((violetLightWavelength / wavelengths.b) ** 4.0) * scatteringStrength;
+  }
+
+  private _updateNumViewRaySamples(numViewRaySamples: number) {
+    this._numViewRaySamples = Math.max(0, Math.min(MAX_SAMPLE_POINTS, numViewRaySamples));
+  }
+
+  private _updateNumSunRaySamples(numSunRaySamples: number) {
+    this._numSunRaySamples = Math.max(0, Math.min(MAX_SAMPLE_POINTS, numSunRaySamples));
+  }
+
+  private _updateExposure(exposure: number) {
+    this._exposure = exposure;
+  }
+
+  public bindExposure(uniform: UniformHandle): void {
+    if (!sync(this, uniform)) {
+      uniform.setUniform1f(this._exposure);
+    }
+  }
+
+  public bindInverseRotationInverseEarthScaleMatrix(uniform: UniformHandle): void {
+    if (!sync(this, uniform)) {
+      this._earthScaleMatrix.multiplyMatrixInverseMatrix(this._inverseEllipsoidRotationMatrix, this._scratchMatrix3d);
+      uniform.setMatrix3(Matrix3.fromMatrix3d(this._scratchMatrix3d));
+    }
+  }
+
+  public bindInverseRotationInverseAtmosphereScaleMatrix(uniform: UniformHandle): void {
+    if (!sync(this, uniform)) {
+      this._atmosphereScaleMatrix.multiplyMatrixInverseMatrix(this._inverseEllipsoidRotationMatrix, this._scratchMatrix3d);
+      uniform.setMatrix3(Matrix3.fromMatrix3d(this._scratchMatrix3d));
+    }
+  }
+
+  public bindOutScatteringIntensity(uniform: UniformHandle): void {
+    if (!sync(this, uniform))
```

Comment:
You've got lots of little `float` and `int` uniforms. We talked about combining them into vectors, arrays, or matrices to reduce the number of calls into WebGL. Did you reject that idea for some reason?

Reply:
Didn't intentionally leave it out, but I did forget about it last week while reworking SkySphereTechnique.
Just pushed a change that replaces several uniforms with a single 4x4 matrix